recipes-ni/ni-base-system-image-tests/files/fs_permissions_diff.py

```diff
@@ -0,0 +1,215 @@
+import argparse
+import csv
+import datetime
+import mmap
+import os
+import pymongo
+import re
+import subprocess
+import sys
+
+class DB:
+    mongo_db_name = 'rtos'
+    mongo_db_collection_name = 'fs_permissions'
+
+    def __init__(self, server, user, password):
+        mongo_client = pymongo.MongoClient("mongodb://{user}:{pwd}@{server}".format(
+            user=args.user, pwd=args.password, server=args.server
+            ),
+            w=1)
+        db = mongo_client[self.mongo_db_name]
+        self.fs_permissions_collection = db[self.mongo_db_collection_name]
+
+    def insert(self, data):
+        return self.fs_permissions_collection.insert_one(data)
+
+    def find(self, query):
+        return self.fs_permissions_collection.find(query)
+
+    def count_documents(self, query):
+        return self.fs_permissions_collection.count_documents(query)
+
+class Logger:
+    logs = []
+    def log(self, log):
+        self.logs.append(log)
+
+    def first_log(self, log):
+        self.logs.insert(0, log)
+
+    def report(self):
+        for log in self.logs:
+            print(log)
+
+def run_cmd(cmd):
+    output = subprocess.check_output(cmd).strip()
+    return output.decode('utf-8')
+
+class OsVersion:
+    def __init__(self):
+        with open('/etc/os-release', 'rb') as file, mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) as mfile:
+            build_id = re.search(br'BUILD_ID=\"((\d+\.\d+).*)\"', mfile)
+            self.full = build_id.group(1).decode()
+            self.major_minor = build_id.group(2).decode()
+            codename = re.search(br'VERSION_CODENAME=\"(.*)\"', mfile)
+            self.codename = codename.group(1).decode()
+
+        decomposed_version = re.match(r'(\d+).(\d+).(\d+)([dabf])(\d+)', self.full)
+        self.major = int(decomposed_version.group(1))
+        self.minor = int(decomposed_version.group(2))
+        self.patch = int(decomposed_version.group(3))
+        self.phase = decomposed_version.group(4)
+        self.build = int(decomposed_version.group(5))
+        self.version_dict = {
+            'major': self.major,
+            'minor': self.minor,
+            'patch': self.patch,
+            'phase': { 'd': 0, 'a': 1, 'b': 2, 'f': 3 }[self.phase],
+            'build': self.build
+        }
+
+fs_manifest_format = '%p\t%M\t%u\t%g\t%l'
+fs_manifest_columns = ['path', 'mode', 'user', 'group', 'link_target']
+def get_fs_manifest():
+    search_dirs = ['/bin', '/boot', '/etc', '/lib', '/lib64', '/sbin', '/usr', '/var']
+    omit_dirs = ['/var/cache', '/var/run', '/var/tmp', '/var/volatile']
```

Comment:
It just occured to me that kernel modules directory `/lib/modules/<kernelversion>` will have a new name on every update. Will this cause false positives? Do we need to omit that directory? I'm not sure if there is any benefit to indexing that directory.

Reply:
We could omit `/lib/modules` if we wanted to. As mentioned in the PR text added and removed files (and by extension, renames like this) are only logged, and won't fail the test. Let me know of your preference.

Reply:
If we keep it, all the files within that directory will be logged right? That'd be too much noise so I'd prefer we not keep it.

Reply:
Makes sense. I'll push that change.